### ISSUE TITLE
[HOLD for payment 2024-11-18] [$250] Settings - The timezone list returns to the top of the list when holding down Tab

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#44245]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.1-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** crowdtest.expensify+test002@gmail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-internal team

## Action Performed:
1. Open the app
2. Go to Account Settings > Profile
3. Open the Timezone > Open time zone list
4. From the first item (Tap on the Tab button, and the time zone will scroll to the top)
5. Keep long pressing a Tab key (~15s)
6. Tap on the Tab button for more time

## Expected Result:
The list will run gradually to the end of the list.

## Actual Result:
The timezone list returns to the top of the list when holding down the Tab. Observation can be seen that the list does not run to the end but gradually moves upward when holding the Tab button for a period of time.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/d1b0e8e1-0ad4-47b3-95ec-f8aa2981d5c5

https://github.com/Expensify/App/assets/115492554/4738993c-4761-42bc-8608-3803f2db43e7

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0198bd49e44880048d</li>
        <li>Upwork Job ID: 1805660378353287866</li>
        <li>Last Price Increase: 2024-07-23</li>
<li>Automatic offers: </li>
<ul>
<li>QichenZhu | Contributor | 103311257</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @lschurr</details>

**Comment by lschurr** [comment]
@allroundexperts Can you review the proposal here?

**Comment by allroundexperts** [comment]
Thanks for your proposal @devguest07. I don't think debouncing is the proper solution here. Your RCA seems to be unconvincing to me. Try pressing tab on GH for example. It would switch focus smoothly.

**Comment by allroundexperts** [comment]
Still looking for better proposals.

**Comment by lschurr** [comment]
Waiting on proposals.

**Comment by lschurr** [comment]
Still waiting on proposals

**Comment by lschurr** [comment]
Same as above.

**Comment by drminh2807** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The timezone list returns to the top of the list when holding down Tab

### What is the root cause of that problem?
- When press `tab`, browser will find next focusable element in the screen, then focus it. ListItem received `onFocus` event then list will scroll to focused item.
- While holding `tab`, the list will scroll rapidly, make items not have enough time to display, then browser will focus to first item as a cycle

### What changes do you think we should make in order to solve the problem?

- Disable `focusable` of ListItem, then handle press `tab` manually like press `arrow-down`

### What alternative solutions did you explore? (Optional)
- Replace SectionList with FlashList to make render time faster, I haven't try it, but it should be a suggestion

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by lschurr** [comment]
Looks like we have a new proposal here @allroundexperts

**Comment by lschurr** [comment]
Bump on this @allroundexperts

**Comment by lschurr** [comment]
Bumped in Slack - https://expensify.slack.com/archives/C01GTK53T8Q/p1721243256819679

**Comment by allroundexperts** [comment]
@drminh2807 Thanks for your proposal. I think that disabling the focus is not fixing the issue at its root. Are we following a similar pattern elsewhere?

**Comment by lschurr** [comment]
Bump on this @drminh2807

**Comment by allroundexperts** [comment]
Still looking for proposals.

**Comment by QichenZhu** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue.
The timezone list often focuses on the text input after holding the Tab key for a long while.

### What is the root cause of that problem?
The text input in the base selection list is set to select its text on focus.

https://github.com/Expensify/App/blob/17a8bbd8e9a5730f7979ba816d6d373d1787bc8f/src/components/SelectionList/BaseSelectionList.tsx#L689

However, the text selection is executed after a timeout, so the text input is selected and re-focused regardless of where the focus is, since the timeout has lower priority than other synchronous code and micro-tasks.

https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/exports/TextInput/index.js#L287-L297
<img width="906" alt="Screenshot 2024-07-24 at 2 38 27 AM" src="https://github.com/user-attachments/assets/6a0742e7-2b1b-4082-aad8-9eaf8b0d09ba">

### What changes do you think we should make in order to solve the problem?
Consider setting the `selectTextOnFocus` attribute of the `TextInput` in [`BaseSelectionList`](https://github.com/Expensify/App/blob/17a8bbd8e9a5730f7979ba816d6d373d1787bc8f/src/components/SelectionList/BaseSelectionList.tsx#L689) to `false` on web and desktop platforms.

### What alternative solutions did you explore? (Optional)
Another option is to check if the text input is still focused before selecting the text in the [`setTimeout`](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/exports/TextInput/index.js#L292-L296) function.

**Comment by lschurr** [comment]
@allroundexperts - Will you check out the new proposal?

**Comment by lschurr** [comment]
Bump on this one @allroundexperts

**Comment by luacmartins** [comment]
I agree that fixing it in `react-native-web` would be ideal.

**Comment by QichenZhu** [comment]
Raised an [issue](https://github.com/necolas/react-native-web/issues/2704) and [PR](https://github.com/necolas/react-native-web/pull/2705) in `react-native-web`. Waiting for their review.

**Comment by lschurr** [comment]
Any update on this one @QichenZhu?

**Comment by QichenZhu** [comment]
My PR is still waiting for `react-native-web` to review. Should we use `patch-package` in the meantime?

**Comment by allroundexperts** [comment]
I don't think so. Let's wait and see if they approve.

**Comment by QichenZhu** [comment]
No updates. Awaiting upstream review.

**Comment by QichenZhu** [comment]
Can we move this to weekly?

**Comment by luacmartins** [comment]
Done

**Comment by lschurr** [comment]
Looks like we're still waiting on the PR review

**Comment by lschurr** [comment]
Looks like all changes on the PR were approved - anything holding us back on this one @QichenZhu @allroundexperts?

**Comment by QichenZhu** [comment]
@lschurr The PR hasn't been approved by a `react-native-web` maintainer yet.

**Comment by QichenZhu** [comment]
Update: The upstream PR https://github.com/necolas/react-native-web/pull/2705 was merged.

**Comment by QichenZhu** [comment]
@allroundexperts The upstream fix has been released in `react-native-web` 0.19.13.

What’s our next step? Should we bump the version now or patch the current version 0.19.12?

cc: @lschurr @luacmartins

**Comment by allroundexperts** [comment]
That's correct @QichenZhu!

**Comment by QichenZhu** [comment]
Thanks @allroundexperts. We got two options:

**Option 1:** Upgrade `react-native-web` to version 0.19.13 and migrate the existing 7 patches.
**Option 2:** Add a new patch to cherry-pick [commit](https://github.com/necolas/react-native-web/commit/e92222446f36a0ec446900fc1b5570c9c6adbd30) to the current version.

What do you think?

**Comment by allroundexperts** [comment]
I think a new patch would be better but lets confirm this from @luacmartins first!

**Comment by luacmartins** [comment]
Asked in Slack https://expensify.slack.com/archives/C01GTK53T8Q/p1729196633652059

**Comment by luacmartins** [comment]
It seems like upgrading `react-native-web` and migrating the patches is the way to go.

**Comment by twisterdotcom** [comment]
Relabelling as Lauren is OOO now.

**Comment by QichenZhu** [comment]
I’d like to request increasing the compensation since we fixed the issue from upstream, upgraded RNWeb to the latest version, migrated existing patches, and made the fix easy to apply to other selection lists. Would $500 sound fair? @luacmartins @isabelastisser

**Comment by isabelastisser** [comment]
@luacmartins, can you please share your opinion about the request above? Thanks!

**Comment by luacmartins** [comment]
That sounds good to me

**Comment by isabelastisser** [comment]
Payment summary:

@allroundexperts requires payment through NewDot Manual Requests $500 for C+ review
@QichenZhu requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/103311257) (Contributor) $500 paid in Upwork.

**Comment by isabelastisser** [comment]
All set!

**Comment by allroundexperts** [comment]
## Checklist
1. This was a bug in the upstream library. As such, we don't have a PR to blame.
2. N/A
3. N/A
4. A regression test would be a good addition.

## Regression test
1. Open the New Expensify web or desktop app.
2. Go to Settings > Profile > Timezone > Timezone.
3. Press the Tab key to focus the first timezone.
4. Hold the Tab key for about 15 seconds.
5. If the interface is unresponsive, wait until it becomes interactive (hover over an element to check). Remember the focused element.
6. Press the Tab key again.

Verify that the element next to the previously focused one is now focused, or the back button is focused if the previously focused element was the last one.

Do we 👍 or 👎 ?

**Comment by JmillsExpensify** [comment]
$500 approved for @allroundexperts
